app/src/main/cpp/skyline/gpu/texture/layout.cpp

```diff
@@ -89,144 +96,339 @@ namespace skyline::gpu::texture {
+
+        size_t pitchBytes{pitchAmount ? pitchAmount : pitchTextureWidthBytes};
+
+        size_t robSize{blockLinearTextureWidthAlignedBytes * robHeight * gobBlockDepth};
+        size_t robPerMob{util::DivideCeil<size_t>(util::DivideCeil<size_t>(blockLinearDimensions.height, formatBlockHeight), robHeight)};
+        size_t blockSize{robHeight * GobWidth * gobBlockDepth};
+
+        u8 *pitchOffset{pitch};
+
+        auto copyTexture{[&]<typename FORMATBPB>() __attribute__((always_inline)) {
+            for (size_t currMob{}; currMob < depthMobCount; ++currMob, blockLinear += robSize * robPerMob) {
+                size_t sliceCount{(currMob + 1) == depthMobCount ? lastMobSliceCount : gobBlockDepth};
+                u64 sliceOffset{};
+                for (size_t slice{}; slice < sliceCount; ++slice, sliceOffset += (GobHeight * GobWidth * gobBlockHeight)) {
+                    u64 robOffset{util::AlignDown(originY, robHeight) * blockLinearTextureWidthAlignedBytes * gobBlockDepth};
+                    for (size_t line{}; line < pitchTextureHeight; ++line, pitchOffset += pitchBytes) {
+                        // XYZ Offset in entire ROBs
+                        if (line && !((originY + line) & (robHeight - 1))) [[unlikely]]
+                            robOffset += robSize;
+                        // Y Offset in entire GOBs in current block
+                        size_t GobYOffset{util::AlignDown((originY + line) & (robHeight - 1), GobHeight) * GobWidth};
+                        // Y Offset inside current GOB
+                        GobYOffset += (((originY + line) & 0x6) << 5) + (((originY + line) & 0x1) << 4);
+
+                        u8 *deSwizzledOffset{pitchOffset};
+                        u8 *swizzledYZOffset{blockLinear + robOffset + GobYOffset + sliceOffset};
+                        u8 *swizzledOffset{};
+
+                        size_t xBytes{originXBytes};
+                        size_t GobXOffset{};
+
+                        size_t blockOffset{util::AlignDown(xBytes, GobWidth) * robHeight * gobBlockDepth};
+
+                        for (size_t pixel{}; pixel < pitchTextureWidth; ++pixel, deSwizzledOffset += formatBpb, xBytes += formatBpb) {
+                            // XYZ Offset in entire blocks in current ROB
+                            if (pixel && !(xBytes & 0x3F)) [[unlikely]]
+                                blockOffset += blockSize;
+
+                            // X Offset inside current GOB
+                            GobXOffset = ((xBytes & 0x20) << 3) + (xBytes & 0xF) + ((xBytes & 0x10) << 1);
+
+                            swizzledOffset = swizzledYZOffset + blockOffset + GobXOffset;
+
+                            if constexpr (BlockLinearToPitch)
+                                *reinterpret_cast<FORMATBPB *>(deSwizzledOffset) = *reinterpret_cast<FORMATBPB *>(swizzledOffset);
```

Comment:
I'd use memcpy here, and template the number of bytes to copy as opposed to the copy type